ocamldep: use `Load_path`

MisterDA opened the pull request:
The referenced `OCAMLCOMP` environment variable was probably a typo, referring to `OCAMLCOMPPARAM`, later renamed to `OCAMLPARAM`. The load path is still reinitialised after the call to `Compenv.readenv`, and the caching is taken care of by `Load_path`.

Entries returned by `Load_path` need to be sorted against the include path(s) to ensure modules are not shadowed.

MisterDA committed:
ocamldep: use Load_path

The referenced `OCAMLCOMP` environment variable was probably a typo,
referring to `OCAMLCOMPPARAM`, later renamed to `OCAMLPARAM`. The load
path is still reinitialised after the call to [Compenv.readenv], and
the caching is taken care of by `Load_path`.

Entries returned by Load_path need to be sorted against the include
path(s) to ensure modules are not shadowed.

gasche commented:
(I wonder if @nojb, our resident Load_path expert, would maybe have the time to look at this.)

nojb commented:
> (I wonder if @nojb, our resident Load_path expert, would maybe have the time to look at this.)

Thanks for the ping. Yes, I will review the PR (hopefully this week).

shindere commented:
Cool! Thanks a lot, @nojb!!

nojb commented on utils/load_path.mli:
```diff
@@ -49,6 +52,9 @@ module Dir : sig
   (** As {!find}, but search also for uncapitalized name, i.e. if name is
       Foo.ml, either /path/Foo.ml or /path/foo.ml may be returned. *)
+
+  val error_occurred : t -> string option
```
This API is error-prone: it is easy to forget to call this function when needed. Instead, the function `create` should return a pair `t * string option` with the error.

MisterDA replied:
My concern was that current consumers of the `create` function simply ignore the error. Is it fine if I use `create path |> fst` in all current call sites?

nojb replied:
I think so. The difference is that the consumers now are _forced_ to ignore the error message (by ignoring the second component of the pair), as opposed to hiding the error in a side effect. Personally I would use pattern matching instead of composing with `fst`, but that's just my taste.

nojb commented on utils/load_path.mli:
```diff
@@ -29,7 +29,10 @@ val remove_dir : string -> unit
-(** Remove all directories *)
+(** Remove all directories and caches. *)
+
+val reset_load_path : unit -> unit
```
This name is confusing (what's the difference between `reset` and `reset_load_path`?).
A better name is needed. Maybe `clear`? (there is a slight analogy with the difference between `Hashtbl.reset` and `Hashtbl.clear`).

MisterDA replied:
Good idea, switched to `clear`, with a bit of documentation.

nojb commented on driver/makedepend.ml:
```diff
@@ -96,23 +78,34 @@ let add_to_synonym_list synonyms suffix =
+  in
+  let names = find Load_path.find and unames = find Load_path.find_normalized in
+  let paths = Load_path.get_paths () in
+  let rec aux current = function
+    | [] -> current
+    | f :: tl ->
+       let dir = Filename.dirname f in
+       if dir = Filename.current_dir_name then Some (dir, Filename.basename f)
+       else match current with
+            | None -> aux (Some (dir, f)) tl
+            | Some (dir', _) ->
+               if is_before dir dir' paths then aux (Some (dir, f)) tl
+               else aux current tl
```
This complicated logic seems to be replicating what `Load_path.find` does. Can't we simply do something along the lines of
```
let names = List.map (fun ext -> name ^ ext) (!mli_synonyms @ !ml_synonyms) in
List.find_map (fun name ->
    match Load_path.find_normalized name with s -> Some s | exception Not_found -> None
  ) names 
```

MisterDA replied:
That was my first idea, but there's a problem with module shadowing. Consider the test
```
testsuite/tests/tool-ocamldep-shadowing
├── a.ml
├── a.reference
├── dir1
│   └── b.ml
└── dir2
    ├── b.mli
    └── c.mli

3 directories, 5 files
```
with the include path `-I dir1 -I dir2` and `a.ml` depending on module `B`. `Load_path` maps file names to complete file paths, keeping the entry with highest priority in the load path. It would associate `b.ml` to `dir1/b.ml` and `b.mli` to `dir2/b.mli`. Now, if we're building possible file names providing module `B`, we'll have `b.mli` first, and then `b.ml`. So, looking up for `B`, we'll find `dir2/b.mli` which is incorrect. `Load_path` doesn't account for module shadowing.

nojb replied:
I need a bit more time to think about this, but if the lookup semantics do not correspond to `Load_path`, why are we trying to use it here? Can we implement the necessary semantics as an alternative "lookup" operation in `Load_path` itself (that would take shadowing into account)?

MisterDA replied:
Well, the code in makedepend is _almost_ a duplicate of `Load_path`, apart from the error reporting and this lookup subtlety, which is why I tried to get makedepend to use `Load_path`.
Sure, I can spend some time on this and try to get `Load_path` to understand shadowing.

MisterDA committed:
Add error reporting to Load_path

Optionally report an error if Sys.readdir failed. Current consumers don't care about the error.

MisterDA committed:
Allow clearing the load path without removing the caches

nojb commented:
@MisterDA Friendly ping after two years: is this still relevant? Otherwise I would prefer to close the PR to better focus on active tasks. Thanks!

MisterDA commented:
Well, maybe there's still an issue in `Load_path` related to module shadowing. I'm not sure if `Load_path` should have different semantics than `ocamldep`. If they should be the same, it makes sense to make `ocamldep` use `Load_path` and have `Load_path` take into account module shadowing, if not, there is not much to do.

I tried rebasing this PR just in case, but I don't understand well the semantics of hidden/visible directories for inclusion. I don't think there is much to be salvaged from this PR.

shindere commented:
Hello,
Apologies for being so silent, the circumstances do not seem to be
aligned fo regular contributions to the compiler, sadly.

I very much think ocamldep and the compiler itslef should agree on
load_path, ideally by using exactly the same code.

I regret I am not able to participate now but I very much hope the
discrepancy will be fixed, one way or another.
